magefile.go

```diff
@@ -15,15 +15,147 @@ import (
+	if err != nil {
+		return err
+	}
+
+	// TODO: Necessary to avoid connection error on Armada server startup.
+	time.Sleep(10 * time.Second)
+	err = sh.Run("docker-compose", "up", "-d", "server", "executor")
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(5 * time.Second)
+	if err := sh.Run("docker", "ps", "-a"); err != nil {
+		return err
+	}
+	sh.Run("docker", "logs", "server")
+	sh.Run("docker", "logs", "executor")
+
+	err = sh.Run("go", "run", "cmd/armadactl/main.go", "create", "queue", "e2e-test-queue")
+	if err != nil {
+		return err
+	}
+	err = sh.Run(
+		"go", "run", "cmd/testsuite/main.go", "test",
+		"--tests", "testsuite/testcases/basic/*",
+		"--junit", "junit.xml",
+	)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func DockerBundle() error {
+	mg.Deps(DockerBundleGoreleaserConfig)
+	return sh.Run(
+		"goreleaser", "release", "--snapshot", "--rm-dist", "-f", ".goreleaser-docker.yml",
+	)
+}
+
+// Write a minimal goreleaser config to .goreleaser-docker.yml
+// containing only the subset of targets in .goreleaser.yaml necessary
+// for building a set of specified Docker images.
+func DockerBundleGoreleaserConfig() error {
+	// Docker targets to build and the build targets necessary to do so.
+	dockerIds := map[string]bool{
+		"bundle": true,
+	}
+	buildIds := map[string]bool{
+		"server":          true,
+		"executor":        true,
+		"binoculars":      true,
+		"lookoutingester": true,
+		"eventingester":   true,
+	}
+
+	goreleaserConfigPath := "./.goreleaser.yml"
+	config := goreleaserConfig.Project{}
+	if err := util.BindJsonOrYaml(goreleaserConfigPath, &config); err != nil {
+		return err
+	}
+
+	dockers := make([]goreleaserConfig.Docker, 0)
+	for _, docker := range config.Dockers {
+		if dockerIds[docker.ID] {
+			dockers = append(dockers, docker)
+		}
+	}
+	if len(dockers) == 0 {
+		return errors.Errorf("%v matched no dockers in %s", dockerIds, goreleaserConfigPath)
+	}
+
+	builds := make([]goreleaserConfig.Build, 0)
+	for _, build := range config.Builds {
+		if buildIds[build.ID] {
+			builds = append(builds, build)
+		}
+	}
+	if len(builds) == 0 {
+		return errors.Errorf("%v matched no builds in %s", buildIds, goreleaserConfigPath)
+	}
+
+	targets := make(map[string]interface{})
+	for _, docker := range dockers {
+		targets[fmt.Sprintf("%s_%s", docker.Goos, docker.Goarch)] = true
+	}
+	for i := range builds {
+		builds[i].Goos = nil
+		builds[i].Goarch = nil
+		builds[i].Targets = maps.Keys(targets)
+	}
+
+	minimalConfig := goreleaserConfig.Project{
+		ProjectName: config.ProjectName,
+		Dist:        config.Dist,
+		GoMod:       config.GoMod,
+		Env:         config.Env,
+		Builds:      builds,
+		Dockers:     dockers,
+	}
+	bytes, err := yaml.Marshal(minimalConfig)
+	if err != nil {
+		return err
+	}
+	f, err := os.Create(".goreleaser-docker.yml")
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if n, err := f.Write(bytes); err != nil {
```

Comment:
I think you can save yourself a couple of lines here by importing `io/ioutil` and then ` ioutil.WriteFile`

arguably we should make a function in common that marshals a struct to yaml and writes to a file, but that's probbaly not worth doing right now.